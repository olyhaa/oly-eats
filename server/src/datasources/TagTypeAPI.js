import { DataSource } from 'apollo-datasource';

class TagTypeAPI extends DataSource {
  constructor({ store }) {
    super();
    this.store = store;
  }

  async getAllTagTypes() {
    const tagTypeArray = await this.store.TagType.findAll();

    return Array.isArray(tagTypeArray)
      ? tagTypeArray.map((tag_type) => this.tagTypeReducer(tag_type))
      : [];
  }

  async addTagType({ label }) {
    const tagType = await this.store.TagType.create({
      label,
    });

    return this.tagTypeMutationReducer({
      success: true,
      tagType,
    });
  }

  async deleteTagType({ id }) {
    const tagType = await this.store.TagType.findByPk(id);
    if (!tagType) {
      return this.tagTypeMutationReducer({
        success: false,
        message: 'ID not found',
      });
    }
    await tagType.destroy();
    return this.tagTypeMutationReducer({
      success: true,
    });
  }

  async updateTagType({ id, label }) {
    let tagType = await this.store.TagType.findByPk(id);
    if (!tagType) {
      return this.tagTypeMutationReducer({
        success: false,
        message: 'ID not found',
      });
    }
    await tagType.update({ label: label });
    tagType = await this.store.TagType.findByPk(id);
    return this.tagTypeMutationReducer({ tagType, success: true });
  }

  tagTypeReducer(tagType) {
    if (!tagType) {
      return null;
    }
    return {
      id: tagType.id,
      label: tagType.label,
    };
  }

  tagTypeMutationReducer({
    success = false,
    message = undefined,
    tagType = null,
  }) {
    return {
      success,
      message,
      tagType: this.tagTypeReducer(tagType),
    };
  }
}

export default TagTypeAPI;
