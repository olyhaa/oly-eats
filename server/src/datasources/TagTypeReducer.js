export const tagTypeReducer = (tagType) => {
  if (!tagType) {
    return null;
  }
  return {
    id: tagType.id,
    label: tagType.label,
    tags: tagType.getTags(),
  };
};

export const tagTypeMutationReducer = ({
  success = false,
  message = undefined,
  tagType = null,
}) => {
  return {
    success,
    message,
    tagType: tagTypeReducer(tagType),
  };
};
