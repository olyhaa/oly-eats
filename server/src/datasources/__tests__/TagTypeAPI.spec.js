import TagTypeAPI from '../TagTypeAPI';
import { tagTypeReducer } from '../TagTypeReducer';

const mockTagTypes = [
  {
    id: '101',
    label: 'category',
    getTags: jest.fn(),
  },
  {
    id: '102',
    label: 'cuisine',
    getTags: jest.fn(),
  },
  {
    id: '103',
    label: 'meal_type',
    getTags: jest.fn(),
  },
];

const mockStore = {
  TagType: {
    create: jest.fn(),
    findAll: jest.fn(),
    findByPk: jest.fn(),
  },
};
const tagTypes = new TagTypeAPI({ store: mockStore });

describe('getAllTagTypes', () => {
  it('returns empty array for empty db', async () => {
    mockStore.TagType.findAll.mockReturnValueOnce([]);
    const response = await tagTypes.getAllTagTypes();
    expect(mockStore.TagType.findAll).toBeCalledWith();
    expect(response).toEqual([]);
  });

  it('returns array of tag types', async () => {
    mockStore.TagType.findAll.mockReturnValueOnce(mockTagTypes);
    mockTagTypes[0].getTags.mockReturnValueOnce([]);
    mockTagTypes[1].getTags.mockReturnValueOnce([]);
    mockTagTypes[2].getTags.mockReturnValueOnce([]);

    const response = await tagTypes.getAllTagTypes();
    expect(mockStore.TagType.findAll).toBeCalledWith();
    expect(response).toMatchSnapshot();
  });
});

describe('addTagType', () => {
  it('calls store creator and returns result', async () => {
    const test_label = 'test label';
    mockStore.TagType.create.mockReturnValueOnce(mockTagTypes[2]);
    mockTagTypes[2].getTags.mockReturnValueOnce([]);

    // check the result of the fn
    const res = await tagTypes.addTagType({ label: test_label });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.TagType.create).toBeCalledWith({
      label: test_label,
    });
  });
});

describe('deleteTagType', () => {
  it('calls store destroy and returns result - id exists', async () => {
    const findPkVal = Object.assign({}, mockTagTypes[2]);
    findPkVal.destroy = jest.fn();
    mockTagTypes[2].getTags.mockReturnValueOnce([]);

    mockStore.TagType.findByPk.mockReturnValueOnce(findPkVal);
    findPkVal.destroy.mockReturnValueOnce('done');

    // check the result of the fn
    const res = await tagTypes.deleteTagType({ id: findPkVal.id });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.TagType.findByPk).toBeCalledWith(findPkVal.id);
    expect(findPkVal.destroy).toBeCalled();
  });

  it('calls store destroy and returns result - id not found', async () => {
    mockStore.TagType.findByPk.mockReturnValueOnce(null);

    // check the result of the fn
    const res = await tagTypes.deleteTagType({ id: 'bad_id' });
    expect(res).toMatchSnapshot();

    expect(mockStore.TagType.findByPk).toBeCalledWith('bad_id');
  });
});

describe('updateTagType', () => {
  it('calls store update with values and returns result - id exists', async () => {
    const findPkVal = Object.assign({}, mockTagTypes[1]);
    const updatedPkVal = Object.assign({}, findPkVal);
    updatedPkVal.label = 'new label';
    mockTagTypes[1].getTags.mockReturnValueOnce([]);

    findPkVal.update = jest.fn();

    mockStore.TagType.findByPk
      .mockReturnValueOnce(findPkVal)
      .mockReturnValueOnce(updatedPkVal);
    findPkVal.update.mockReturnValueOnce(updatedPkVal);
    const new_label = 'new label';

    // check the result of the fn
    const res = await tagTypes.updateTagType({
      id: findPkVal.id,
      label: new_label,
    });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.TagType.findByPk).toBeCalledWith(findPkVal.id);
    expect(findPkVal.update).toBeCalledWith({ label: new_label });
  });

  it('calls store update and returns result - id not found', async () => {
    mockStore.TagType.findByPk.mockReturnValueOnce(null);
    const bad_id = 'bad_id';

    // check the result of the fn
    const res = await tagTypes.updateTagType({
      id: bad_id,
      label: 'new label',
    });
    expect(res).toMatchSnapshot();

    // make sure store is called properly
    expect(mockStore.TagType.findByPk).toBeCalledWith(bad_id);
  });
});
