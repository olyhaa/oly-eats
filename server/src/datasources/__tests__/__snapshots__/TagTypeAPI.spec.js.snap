// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTagType calls store creator and returns result 1`] = `
Object {
  "message": undefined,
  "success": true,
  "tagType": Object {
    "id": "103",
    "label": "meal_type",
    "tags": Array [],
  },
}
`;

exports[`deleteTagType calls store destroy and returns result - id exists 1`] = `
Object {
  "message": undefined,
  "success": true,
  "tagType": null,
}
`;

exports[`deleteTagType calls store destroy and returns result - id not found 1`] = `
Object {
  "message": "ID not found",
  "success": false,
  "tagType": null,
}
`;

exports[`getAllTagTypes returns array of tag types 1`] = `
Array [
  Object {
    "id": "101",
    "label": "category",
    "tags": Array [],
  },
  Object {
    "id": "102",
    "label": "cuisine",
    "tags": Array [],
  },
  Object {
    "id": "103",
    "label": "meal_type",
    "tags": Array [],
  },
]
`;

exports[`updateTagType calls store update and returns result - id not found 1`] = `
Object {
  "message": "ID not found",
  "success": false,
  "tagType": null,
}
`;

exports[`updateTagType calls store update with values and returns result - id exists 1`] = `
Object {
  "message": undefined,
  "success": true,
  "tagType": Object {
    "id": "102",
    "label": "new label",
    "tags": Array [],
  },
}
`;
